### PR TITLE
CP-11007: Update VDI.snapshot_of links after VM.revert

### PR DESCRIPTION
**Pull request by ravippandey**
This fix also includes the rebased fix of CP-10974

Signed-off-by: Ravi Pandey <ravi.pandey@citrix.com>

### CODE REVIEW

**Comment by johnelse** [comment]
@ravip please could you split this up into two commits? One for the CA-118466 fix (if you're lucky this will cherry-pick cleanly from the original pull request) and one for the new fix.

**Comment by ravippandey** [comment]
@johnelse yes sure

**Comment by ravippandey** [comment]
@johnelse Done. Sorry, I didnt realise that siddharth's commit was getting hidden earlier. Thanks

**Comment by johnelse** [comment]
@ravippandey I know it wasn't originally your commit message, but please rewrite the first message so that the first line is shorter and the second line is blank, as per [this](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) guide. Extra information can go on the third line onwards.

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 1]
```diff
@@ -289,9 +299,19 @@ let update_vifs_vbds_and_vgpus ~__context ~snapshot ~vm =
 
 		debug "Cloning the snapshoted disks";
 		let driver_params = Xapi_vm_clone.make_driver_params () in
-		let cloned_disks = Xapi_vm_clone.safe_clone_disks rpc session_id Xapi_vm_clone.Disk_op_clone ~__context snap_vbds driver_params in
+		let cloned_disks = Xapi_vm_clone.safe_clone_disks rpc session_id Xapi_vm_clone.Disk_op_clone ~__context snap_vbds_without_cd driver_params in
 		TaskHelper.set_progress ~__context 0.5;
 
+		debug "Updating the snapshot_of fields for relevant VDIs";
+		let original_to_newvdi_map = List.map2 (fun oldvdi (_, newvdi, _) -> (oldvdi,newvdi)) vdis_snap_of cloned_disks in
```
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: Renaming `original_to_newvdi_map` to `oldvdi_to_newvdi_map` would be more consistent.

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 2]
```diff
@@ -289,9 +299,19 @@ let update_vifs_vbds_and_vgpus ~__context ~snapshot ~vm =
...
+		List.iter (fun (vdi, snaps) ->
```
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: `vdi` would make more sense if it was called `oldvdi`

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 3]
```diff
@@ -289,9 +299,19 @@ let update_vifs_vbds_and_vgpus ~__context ~snapshot ~vm =
...
+			let newvdi = List.assoc vdi original_to_newvdi_map in
+			List.iter (fun s -> Db.VDI.set_snapshot_of ~__context ~self:s ~value:newvdi) snaps
+		) vdi_to_snapshots_map;
+
+		debug "Updating the snapshot_of fields for snapshot VDIs";
+		List.iter2 (fun snapvdi (_, newvdi, _) -> Db.VDI.set_snapshot_of ~__context ~self:snapvdi ~value:newvdi ) snap_vdis cloned_disks;
```
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: Sorry if I've misunderstood, but why are these two lines needed? Haven't we already updated the `snapshot_of` field of each snapshot of each VDI at this point?

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 4]
```diff
@@ -289,9 +299,16 @@ let update_vifs_vbds_and_vgpus ~__context ~snapshot ~vm =
 
 		debug "Cloning the snapshoted disks";
 		let driver_params = Xapi_vm_clone.make_driver_params () in
-		let cloned_disks = Xapi_vm_clone.safe_clone_disks rpc session_id Xapi_vm_clone.Disk_op_clone ~__context snap_vbds driver_params in
+		let cloned_disks = Xapi_vm_clone.safe_clone_disks rpc session_id Xapi_vm_clone.Disk_op_clone ~__context snap_vbds_without_cd driver_params in
 		TaskHelper.set_progress ~__context 0.5;
 
+		debug "Updating the snapshot_of fields for relevant VDIs";
+		let oldvdi_to_newvdi_map = List.map2 (fun oldvdi (_, newvdi, _) -> (oldvdi,newvdi)) vdis_snap_of cloned_disks in
+		List.iter (fun (vdi, snaps) ->
```
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: I still think `vdi` should be `oldvdi`

File: ocaml/xapi/xapi_vm_snapshot.ml [hunk 5]
```diff
@@ -269,16 +269,26 @@ let safe_destroy_vdi ~__context ~rpc ~session_id vdi =
...
-	let vm_VDIs = List.map (fun vbd -> Db.VBD.get_VDI __context vbd) vm_VBDs in
+	(* Filter VBDs to ensure that we don't read empty CDROMs *)
+	let vbds_without_cd = List.filter (fun vbd -> Db.VBD.get_type ~__context ~self:vbd <> `CD) vm_VBDs in
+	let vm_VDIs = List.map (fun vbd -> Db.VBD.get_VDI ~__context ~self:vbd) vbds_without_cd in
+	let vm_VDIs = List.filter (fun vdi -> List.mem vdi vdis_snap_of) vm_VDIs in
 	let vm_VIFs = Db.VM.get_VIFs ~__context ~self:vm in
 	let vm_VGPUs = Db.VM.get_VGPUs ~__context ~self:vm in
 	let vm_suspend_VDI = Db.VM.get_suspend_VDI ~__context ~self:vm in
 
+	let vdi_to_snapshots_map = List.map (fun vdi -> let snaps = Db.VDI.get_snapshots ~__context ~self:vdi in
```
Comment (johnelse) [inline ocaml/xapi/xapi_vm_snapshot.ml]: I think this would be clearer if it was called `oldvdi_to_snapshots_map`

**Comment by johnelse** [comment]
I think this looks good now, thanks for all the changes @ravippandey ! Please go ahead and squash the three CP-11007 commits into one, then this can be merged.

**Comment by johnelse** [comment]
Ready for merge once the trunk-ring3 BVT is verified to be passing.

**Pull request merged**